4-General-code-optimization/README.md

```diff
@@ -0,0 +1,121 @@
+# How to Optimize GPU utilization
+
+This tutorial covers the basic concepts for achieving maximum GPU compute utilization during deep learning training.
+This can be achieved in a number of ways, and can be broadly broken down into the following sections:
+
+## 1. Make the most of the pytorch dataloader:
+
+The built-in Pytorch
+dataloader, [torch.utils.data.DataLoader](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader), has a
+number of arguments you can define in order to improve the throughput of data being loaded.
+
+This includes:
+
+- Setting `batch_size` to as high as possible before reaching an OOM error (VRAM too full to fit all data). This
+  maximizes the time the GPU spends on performing computations, before having to wait for more data to be loaded.
+- Setting `num_workers` > 0 to use multiple cpu cores for loading data. This is to minimize the time spent by the GPU
+  waiting idly for the CPU to transfer data to it for processing.
+- Raising value of `prefetch_factor`.
+- Setting `pin_memory=True`.
+- Setting `persistent_workers=True`
+
+A good default dataloader may look like this:
+
+```python
+my_dataloader = torch.utils.data.DataLoader(my_dataset,
+                                            batch_size=my_max_batchsize,
+                                            num_workers=my_max_cpu_count,
```

Comment:
@david-stojanovski max cpu count is a little dangerous here. When `--cpu-limit` is undefined, it defaults to the maximum available cpu cores.

Reply:
that's a good point, i'll add a warning comment next to it